stan/math/prim/mat/fun/cov_periodic.hpp

```diff
@@ -0,0 +1,144 @@
+
+      Eigen::Matrix<typename stan::return_type<T_x, T_sigma, T_l, T_p>::type,
+                    Eigen::Dynamic, Eigen::Dynamic> cov(x.size(), x.size());
+
+      size_t x_size = x.size();
+      if (x_size == 0)
+        return cov;
+
+      T_sigma sigma_sq = square(sigma);
+      T_l neg_two_inv_l_sq = -2.0 * inv_square(l);
+      T_p pi_div_p = M_PI * inv(p);
+
+      for (size_t j = 0; j < x_size; ++j) {
+        cov(j, j) = sigma_sq;
+        for (size_t i = j + 1; i < x_size; ++i) {
+          cov(i, j) = sigma_sq * exp(square(sin(pi_div_p * distance(x[i], x[j])))
+                                     * neg_two_inv_l_sq);
+          cov(j, i) = cov(i, j);
+        }
+      }
+      return cov;
+    }
+
+    /**
+     * Returns a periodic covariance matrix \f$ \mathbf{K} \f$ using inputs
+     * \f$ \mathbf{X}_1 \f$ and \f$ \mathbf{X}_2 \f$.
+     * The elements of \f$ \mathbf{K} \f$ are defined as
+     * \f$ \mathbf{K}_{ij} = k(\mathbf{X}_{1_i},\mathbf{X}_{2_j}), \f$ where
+     * \f$ \mathbf{X}_{1_i} \f$ and \f$ \mathbf{X}_{2_j} \f$  are the \f$i\f$-th and \f$j\f$-th rows of
+     * \f$ \mathbf{X}_1 \f$ and \f$ \mathbf{X}_2 \f$ and \n
+     * \f$ k(\mathbf{x},\mathbf{x}^\prime) =
+     * \sigma^2 \exp\left(-\frac{2\sin^2(\pi |\mathbf{x}-\mathbf{x}^\prime|/p)}{\ell^2}\right), \f$
+     * \n
+     * where \f$ \sigma^2 \f$, \f$ \ell \f$ and \f$ p \f$ are the signal variance, length-scale and period.
+     *
+     * @tparam T_x1 type of std::vector elements of x1
+     *   T_x1 can be a scalar, an Eigen::Vector, or an Eigen::RowVector.
+     * @tparam T_x2 type of std::vector elements of x2
+     *   T_x2 can be a scalar, an Eigen::Vector, or an Eigen::RowVector.
+     * @tparam T_sigma type of sigma
+     * @tparam T_l type of length-scale
+     * @tparam T_p type of period
+     *
+     * @param x1 std::vector of first input elements
+     * @param x2 std::vector of second input elements.
+     *   This function assumes that all the elements of x1 and x2 have the same sizes.
+     * @param sigma standard deviation of the signal
+     * @param l length-scale
+     * @param p period
+     * @return periodic covariance matrix
+     * @throw std::domain_error if sigma <= 0, l <= 0, p <= 0 ,
+     *   x1 or x2 is nan or infinite
+     */
+    template<typename T_x1, typename T_x2, typename T_sigma, typename T_l, typename T_p>
+    inline typename
+    Eigen::Matrix<typename stan::return_type<T_x1, T_x2, T_sigma, T_l, T_p>::type,
+                  Eigen::Dynamic, Eigen::Dynamic>
+    cov_periodic(const std::vector<T_x1>& x1, const std::vector<T_x2>& x2,
+                 const T_sigma& sigma, const T_l& l, const T_p& p) {
+      using std::exp;
+      check_positive("cov_periodic", "signal standard deviation", sigma);
+      check_positive("cov_periodic", "length-scale", l);
+      check_positive("cov_periodic", "period", p);
+      for (size_t n = 0; n < x1.size(); ++n)
+        check_not_nan("cov_periodic", "x1", x1[n]);
```

Comment:
This should say something like "element of x1", because it's not `x1` itself that is NaN (same for `x2`).  We really need to get the vectorized versions of all these in place, because doing it this way provides poor error messages that don't give you an index.

Reply:
Done. Didn't create separate object for the text as for "cov_periodic", although some allocations could be saved if x is vector.